internal/pkg/cli/job_delete.go

```diff
@@ -164,6 +234,110 @@ func (o *deleteJobOpts) askJobName() error {
 	return nil
 }
 
+func (o *deleteJobOpts) appEnvironments() error {
+	if o.envName != "" {
+		env, err := o.targetEnv()
+		if err != nil {
+			return err
+		}
+		o.environments = append(o.environments, env)
+	} else {
+		envs, err := o.store.ListEnvironments(o.appName)
+		if err != nil {
+			return fmt.Errorf("list environments: %w", err)
+		}
+		o.environments = envs
+	}
+	return nil
+}
+
+func (o *deleteJobOpts) deleteStacks() error {
+	for _, env := range o.environments {
+		sess, err := o.sess.FromRole(env.ManagerRoleARN, env.Region)
+		if err != nil {
+			return err
+		}
+
+		cfClient := o.getJobCFN(sess)
+		o.spinner.Start(fmt.Sprintf(fmtJobDeleteStart, o.name, env.Name))
+		if err := cfClient.DeleteJob(deploy.DeleteJobInput{
+			Name:    o.name,
+			EnvName: env.Name,
+			AppName: o.appName,
+		}); err != nil {
+			o.spinner.Stop(log.Serrorf(fmtJobDeleteFailed, o.name, env.Name, err))
+			return err
+		}
+		o.spinner.Stop(log.Ssuccessf(fmtJobDeleteComplete, o.name, env.Name))
+	}
+	return nil
+}
+
+func (o *deleteJobOpts) needsAppCleanup() bool {
+	// Only remove a job from the app if
+	// we're removing it from every environment.
+	// If we're just removing the job from one
+	// env, we keep the app configuration.
+	return o.envName == ""
+}
+
+// This is to make mocking easier in unit tests
+func (o *deleteJobOpts) emptyECRRepos() error {
```

Comment:
This is a similar feedback to the "job ls" command where so much of the code is the same between both "job delete" and "svc delete".

Maybe one thing that we can start doing is moving the methods that are absolutely identical to a separate `"delete"` package so that we don't duplicate the methods and have multiple places that we need to fix in case of a bug:

```go
// in a new cli/delete package
package delete

type EmptyECRConfig struct {
   App string
   Environments []*config.Environment
   Workload string
   Sess *aws.Session

   getECRClient func(sess *aws.Session) imageRemover
}

func (c *EmptyECRConfig) EmptyRepos() error {
   if c.getECRClient == nil {
        c.getECRClient = func(session *aws.Session) imageRemover {
		return ecr.New(session)
	}
   }
   // all the common logic
}
```

Reply:
I'll work on this in a separate PR! In the meantime: https://github.com/aws/copilot-cli/issues/1477